// htmltocanvas-plus/dist/lib/dom/element-container.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ElementContainer = void 0;
var index_1 = require("../css/index");
var bounds_1 = require("../css/layout/bounds");
var node_parser_1 = require("./node-parser");
var debugger_1 = require("../core/debugger");
var ElementContainer = /** @class */ (function () {
    function ElementContainer(context, element) {
        this.context = context;
        this.textNodes = [];
        this.elements = [];
        this.flags = 0;
        if (debugger_1.isDebugging(element, 3 /* PARSE */)) {
            debugger;
        }
        var elementStyle = __assign({}, window.getComputedStyle(element, null));
        if (this.isBackgroundClip(elementStyle)) {
            elementStyle.backgroundColor = 'transparent';
            elementStyle.backgroundImage = 'none';
        }
        this.styles = new index_1.CSSParsedDeclaration(context, elementStyle);
        if (node_parser_1.isHTMLElementNode(element)) {
            if (this.styles.animationDuration.some(function (duration) { return duration > 0; })) {
                element.style.animationDuration = '0s';
            }
            if (this.styles.transform !== null) {
                // getBoundingClientRect takes transforms into account
                element.style.transform = 'none';
            }
        }
        this.bounds = bounds_1.parseBounds(this.context, element);
        if (debugger_1.isDebugging(element, 4 /* RENDER */)) {
            this.flags |= 16 /* DEBUG_RENDER */;
        }
    }
    ElementContainer.prototype.isBackgroundClip = function (style) {
        var isHave = style.webkitBackgroundClip === 'text' || style.backgroundClip === 'text';
        return isHave && (CSS.supports('-webkit-background-clip', 'text') || CSS.supports('background-clip', 'text'));
    };
    return ElementContainer;
}());
exports.ElementContainer = ElementContainer;
